File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        result = int(1e9)

        for i in range(n):
            lo, hi = i + 1, n - 1

            while lo < hi:
                chk = nums[i] + nums[lo] + nums[hi]
                if abs(target - chk) < abs(result - target):
                    result = chk
                if chk > target:
                    hi -= 1
                elif chk < target:
                    lo += 1
                else:
                    return target

        return result

File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;
        int result = Integer.MAX_VALUE;

        for (int i = 0; i < n; i++) {
            int lo = i + 1;
            int hi = n - 1;

            while (lo < hi) {
                int chk = nums[i] + nums[lo] + nums[hi];
                if (Math.abs(chk - target) < Math.abs(result - target)) {
                    result = chk;
                }
                if (chk > target) {
                    hi--;
                } else if (chk < target) {
                    lo++;
                } else {
                    return target;
                }
            }
        }

        return result;
    }
}